src/box/lua/experimental/connpool.lua

```diff
@@ -92,7 +92,70 @@ local function get_candidates(role_name, opts)
     return candidates
 end
 
+local function get_connection(role_name, opts)
+    checks('string', {
+        labels = '?table',
+        prefer_local = '?boolean',
+    })
+    opts = opts or {}
+
+    local candidates_opts = {
+        labels = opts.labels
+    }
+    local candidates = get_candidates(role_name, candidates_opts)
+    if next(candidates) == nil then
+        if opts.labels ~= nil then
+            local msg = 'No instances with role %q and labels %s are available'
+            error(msg:format(role_name, table.concat(opts.labels)), 0)
+        end
+        local msg = 'No instances with role %q are available'
+        error(msg:format(role_name), 0)
+    end
+
+    if opts.prefer_local == nil or opts.prefer_local == true then
+        local name = box.info.name
+        for n, candidate_name in pairs(candidates) do
+            if candidate_name == name then
+                table.remove(candidates, n)
+                local candidate_opts = {wait_connected = false}
+                local conn = connect_using_config(name, candidate_opts)
+                if conn:wait_connected() then
+                    return conn
+                end
+                break
+            end
+        end
+    end
+
+    while #candidates > 0 do
+        local n = math.random(#candidates)
+        local instance_name = table.remove(candidates, n)
+        local candidate_opts = {wait_connected = false}
+        local conn = connect_using_config(instance_name, candidate_opts)
+        if conn:wait_connected() then
+            return conn
+        end
```

Comment:
I would let it fail if it doesn't became connected within the given/default connect timeout. It sounds better than an infinite waiting.

Reply:
Or `conn:wait_connected()` is not infinite?

BTW, if it is infinite, there is no point in the `while` loop here.